### PR TITLE
Add type declarations

### PR DESCRIPTION
**Pull request by unindented**

### CODE REVIEW

**Comment by unindented** [comment]
@blakeembrey could you please review?

**Comment by blakeembrey** [comment]
@unindented Just remember that TypeScript is going to do structural type checking, so some of those values already appear to be assignable to each other.

File: typings.d.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
+interface EntitySchemaOptions {
```
Comment (blakeembrey) [inline typings.d.ts]: You will probably want to export these types.

Comment (unindented) [inline typings.d.ts]: All types that are used by public APIs should be exported? So I should export everything in this file, right?

**Comment by unindented** [comment]
@blakeembrey Forgive my ignorance, I'm very new to type systems. Could you give me an example to illustrate the fact that "some of those values already appear to be assignable to each other", and what consequences it has?

**Comment by unindented** [comment]
@gaearon What are your thoughts on adding type definitions to the repo? I have an attempt at TypeScript definitions here, and could work on a `.flowconfig` file...

**Comment by blakeembrey** [comment]
@unindented This is the basic gist: http://www.typescriptlang.org/play/index.html#src=class%20Foo%20%7B%7D%0A%0Aclass%20Bar%20%7B%7D%0A%0Afunction%20x%20(foo%3A%20Foo)%20%7B%7D%0A%0Ax(new%20Bar())%0Ax('test')%0Ax(10)%0Ax(Foo)

Since all the interfaces are structurally the same, you aren't actually doing any type checking.

File: typings.d.ts (hunk 1 above)
Comment (blakeembrey) [inline typings.d.ts]: Pretty much, export everything.

**Comment by unindented** [comment]
@blakeembrey Aha, understood. Adding empty interfaces is almost equivalent to `any`, so my union type was useless.

I've updated the PR with proper types.

**Comment by unindented** [comment]
@paularmstrong @gaearon do you think the type definitions belong to this repo? There are ways to host the type definitions in a separate repo, but then we wouldn't be able to reference them in the `package.json`, and it would add some friction for users.

**Comment by paularmstrong** [comment]
I don't see why normalizr shouldn't include its own type declarations. However, is there a way to have an automated test ensure that these are updated when necessary?

**Comment by unindented** [comment]
Umm, I don't know the best way to ensure type declarations are up-to-date. @blakeembrey what do you suggest?

**Comment by gaearon** [comment]
>However, is there a way to have an automated test ensure that these are updated when necessary?

If we add them, we should add tests for them. 
I’m also ignorant of the process, here’s what community did with Redux:

https://github.com/reactjs/redux/pull/1413
https://github.com/reactjs/redux/pull/1526
https://github.com/reactjs/redux/pull/1537
https://github.com/reactjs/redux/pull/1545

These include tests.

**Comment by unindented** [comment]
Ok, I'll follow the same approach.

**Comment by blakeembrey** [comment]
Yes, that's what we usually do too. You can have some valid code and make sure it compiles validly with `tsc`, or even run some of the test suite with TypeScript.

**Comment by unindented** [comment]
Hey @paularmstrong, I finally got around to this.

I've updated the definitions with the latest changes that were made (`getMeta`, etc.) and added tests that cover all signatures.

Let me know what you think!

**Comment by paularmstrong** [comment]
Thanks @unindented. I'll merge this soon for v2.2.0

**Comment by unindented** [comment]
@paularmstrong any planned date for `v2.2.0`?

**Comment by paularmstrong** [comment]
Been on vacation. Maybe can shoot for late next week. If there's a strong request, though, I could defer the docs/examples/tests updates that we need.

**Comment by unindented** [comment]
Tests are already in this PR. Other projects, like `redux`, assume people who are using TypeScript know how to consume types, so they don't add specific docs or examples.

**Comment by paularmstrong** [comment]
@unindented I mean #105, #106, and #107

**Pull request closed without merging**